README.md
cloud_haskell_example

CHANGELOG.md
# Revision history for distr

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


app/Main copy.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TupleSections #-}

import Control.Monad (forM, forM_)
import Data.Typeable
import Control.Concurrent (threadDelay)
import Control.Monad (forever)
import Control.Distributed.Process
import Control.Distributed.Process.Closure
import Control.Distributed.Process.Node
import Network.Transport.TCP (createTransport, defaultTCPParameters, defaultTCPAddr)


data Result a = Result a deriving Show

showResult :: Result a -> Process ()
showResult a = say $ show a

addNumbers :: Int -> Int -> Process Int
addNumbers x y = return (x + y)

stringTransform :: String -> Process String
stringTransform str = return (reverse str)

stringTransformWrapper :: Int -> Int -> Process String
stringTransformWrapper one two = stringTransform $ show one <> show two

remotable ['addNumbers, 'stringTransformWrapper]

myRemoteTable :: RemoteTable
myRemoteTable = Main.__remoteTable initRemoteTable

master :: Process ()
master = do
  

  -- Get information about available slave nodes
  slaves <- findSlaves

  -- Send tasks to the slave nodes
  spawn  slaves


findSlaves :: Process [NodeId]
findSlaves = do
  -- Discover and list available slave nodes (you may need to replace these with actual slave node addresses)
  -- For simplicity, we assume that slave nodes are running on the same machine.
  return $ map NodeId ["localhost:10501", "localhost:10502"]

sendTasks :: [NodeId] -> Process ()
sendTasks slaves = do
  -- Distribute tasks to slaves
  let tasks = [(1, 2), (3, 4)] -- Tasks to distribute
  let actions = cycle $ [addNumbers, stringTransformWrapper]
  results <- forM (zip tasks slaves actions) $ \(task, slave, action) -> do
    pid <- spawn slave $ $(mkClosure 'action) task
    result <- expect
    result
  forM_ results showResult 
  return ()

main :: IO ()
main = do
  Right transport <- flip createTransport defaultTCPParameters $ defaultTCPAddr "localhost" "10500"
  node <- newLocalNode transport myRemoteTable
  runProcess node master


{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TupleSections #-}

import qualified Data.ByteString.Char8 as BS
import Control.Monad (forM, forM_)
import Data.Typeable
import Control.Concurrent (threadDelay)
import Control.Monad (forever)
import Control.Distributed.Process
import Control.Distributed.Process.Closure
import Control.Distributed.Process.Node
import Network.Transport.TCP (createTransport, defaultTCPParameters, defaultTCPAddr)
import Network.Transport


data Message = StringMsg (String, String) |
               NumberMsg (Int, Int)

data (Show a) => Result a = Result { _getResult :: a }
instance (Show a) => Show (Result a) where
  show a = show $ _getResult a

showResult :: (Show a) => Result a -> Process ()
showResult a = say $ show a

addNumbers :: (Int, Int) -> Process ()
addNumbers (x, y) = do 
  let a = (x + y)
  say $ show a

stringTransform :: String -> Process ()
stringTransform str = do 
  let a = reverse str
  say a

stringTransformWrapper :: (Int, Int) -> Process ()
stringTransformWrapper (one, two) = stringTransform $ show one <> show two

remotable ['addNumbers, 'stringTransformWrapper]

myRemoteTable :: RemoteTable
myRemoteTable = Main.__remoteTable initRemoteTable

master :: Process ()
master = do

  -- Get information about available slave nodes
  slaves <- findSlaves

  -- Send tasks to the slave nodes
  sendTasks slaves


findSlaves :: Process [NodeId]
findSlaves = do
  -- Discover and list available slave nodes (you may need to replace these with actual slave node addresses)
  -- For simplicity, we assume that slave nodes are running on the same machine.
  return $ map NodeId [EndPointAddress (BS.pack "localhost:10501"), EndPointAddress ( BS.pack "localhost:10502")]

sendTasks :: [NodeId] -> Process ()
sendTasks slaves = do
  -- Distribute tasks to slaves
  let task1 = [(1, 2), (3, 4)] :: [(Int, Int)] -- Tasks to distribute
  let task2 = [(1, 2), (3, 4)] :: [(Int, Int)] -- Tasks to distribute

  results1 <- forM (zip task1 slaves) $ \(task, slave) -> do
    pid <- spawn slave $ $(mkClosure 'addNumbers) task
    expect
  forM_ results1 showResult 

  results2 <- forM (zip task2 slaves) $ \(task, slave) -> do
    pid <- spawn slave $ $(mkClosure 'stringTransformWrapper) task
    expect
  forM_ results2 showResult 
  return ()

main :: IO ()
main = do
  Right transport <- flip createTransport defaultTCPParameters $ defaultTCPAddr "localhost" "10500"
  node <- newLocalNode transport myRemoteTable
  runProcess node master


app/Main ping.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}


import Text.Printf
import qualified Data.ByteString.Char8 as BS
import System.Environment (getArgs)
import GHC.Generics
import Control.Monad (forM, forM_)
import Data.Typeable
import Data.Binary (Binary)
import Control.Concurrent (threadDelay)
import Control.Monad (forever)
import Control.Distributed.Process
import Control.Distributed.Process.Closure
import Control.Distributed.Process.Node
import Control.Distributed.Process.Serializable
import Network.Transport.TCP (createTransport, defaultTCPParameters, defaultTCPAddr)
import Network.Transport (Transport, EndPointAddress(..))
import Control.Concurrent (forkIO)

type Host = String
type Port = String
getTransport :: Host -> Port -> IO Transport
getTransport host port = do
  Right transport <- flip createTransport defaultTCPParameters $ defaultTCPAddr host port
  return transport

replyBack :: (ProcessId, String) -> Process ()
replyBack (sender, msg) = send sender msg

logMessage :: String -> Process ()
logMessage msg = say $ "handling " ++ msg



data MyMessage = Ping ProcessId
             | Pong ProcessId
  deriving (Typeable, Generic)          -- 1

instance Binary MyMessage  

pingServer :: Process ()
pingServer = do
  Ping from <- expect                              -- 1
  say $ printf "ping received from %s" (show from) -- 2
  mypid <- getSelfPid                              -- 3
  send from (Pong mypid) 

remotable ['pingServer]

myRemoteTable :: RemoteTable
myRemoteTable = Main.__remoteTable initRemoteTable

master :: Process ()
master = do
  node <- getSelfNode                               -- 1

  say $ printf "spawning on %s" (show node)
  pid <- spawn node $(mkStaticClosure 'pingServer)  -- 2

  mypid <- getSelfPid                               -- 3
  say $ printf "sending ping to %s" (show pid)
  send pid (Ping mypid)                             -- 4

  Pong _ <- expect                                  -- 5
  say "pong."

  terminate                                         -- 6


main :: IO ()
main = do
  master_transport <- getTransport "localhost" "10500"
  node <- newLocalNode master_transport myRemoteTable
  runProcess node $ master
-- main :: IO ()
-- main = do
--   args <- getArgs

--   case args of
--     ["master", host, port] -> do
--       backend <- initializeBackend host port myRemoteTable
--       startMaster backend (master backend)
--     ["slave", host, port] -> do
--       backend <- initializeBackend host port myRemoteTable
--       startSlave backend


-- main :: IO ()
-- main = do
--   args <- getArgs

--   case args of
--     ["master", host, port] -> do
--         master_transport <- getTransport host port
--         node <- newLocalNode master_transport myRemoteTable
--         runProcess node $ master
--     ["slave", host, port] -> do
--         master_transport <- getTransport host port
--         node <- newLocalNode master_transport myRemoteTable
--         runProcess node $ (receiveWait[match logMessage, match replyBack])


-- main :: IO ()
-- main = do
--   print "start"
--   master_transport <- getTransport "localhost" "10500"
--   slave1_transport <- getTransport "localhost" "10501"
--   slave2_transport <- getTransport "localhost" "10502"
--   print "start"
  
--   slave_local_nodes <- flip traverse [slave1_transport, slave2_transport] (\tr -> newLocalNode tr myRemoteTable)
--   print "start"
--   let slave_nodes = map localNodeId slave_local_nodes
--   print "start"
--   slaves_process <- forM slave_local_nodes (\tr -> 
--     forkProcess tr (receiveWait[]))
--   print "slave in work"
--   node <- newLocalNode master_transport myRemoteTable
--   runProcess node $ master



app/Main_both.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}


import qualified Data.ByteString.Char8 as BS
import GHC.Generics
import Control.Monad (forM, forM_)
import Data.Typeable
import Data.Binary (Binary)
import Control.Concurrent (threadDelay)
import Control.Monad (forever)
import Control.Distributed.Process
import Control.Distributed.Process.Closure
import Control.Distributed.Process.Node
import Control.Distributed.Process.Serializable
import Network.Transport.TCP (createTransport, defaultTCPParameters, defaultTCPAddr)
import Network.Transport (Transport, EndPointAddress(..))
import Control.Concurrent (forkIO)


data MyMessage = StringMsg (String, String) |
               IntMsg (Int, Int) |
               QueryMsg ProcessId |
               ResultMsg (String, Int) |
               TerminateMsg deriving (Generic, Binary)

data (Show a) => Result a = Result { _getResult :: a }
instance (Show a) => Show (Result a) where
  show a = show $ _getResult a

showResult :: (Show a) => Result a -> Process ()
showResult a = say $ show a

addNumbers :: (Int, Int) -> Int
addNumbers (x, y) = (x + y)

stringTransform :: String -> String
stringTransform str = reverse str

stringTransformWrapper :: (String, String) -> String
stringTransformWrapper (one, two) = stringTransform $ show one <> show two

slaveProcess :: (String, Int) -> Process ()
slaveProcess (s, i) = do
  say $ s <> show i
  (s', i') <- receiveWait[match commandCatch]
  say $ s' <> show i'
  slaveProcess (s', i')
  where
    commandCatch (QueryMsg pid) = do
      send pid (s, i)
      return (s, i)
    commandCatch (StringMsg (s1, s2)) = return ((s <> (stringTransformWrapper (s1, s2))), i)
    commandCatch (IntMsg (i1, i2)) = return (s, (i + (addNumbers (i1, i2))))
    commandCatch TerminateMsg = terminate

query :: ProcessId -> ProcessId -> Process ()
query cpid mypid = do
  send cpid (QueryMsg mypid)


remotable ['slaveProcess]

myRemoteTable :: RemoteTable
myRemoteTable = Main.__remoteTable initRemoteTable

master :: Process ()
master = do
  say "master started"
  -- Get information about available slave nodes
  slaves <- findSlaves
  say "master started"

  processes <- getProcess slaves
  say "master started"
  mypid <- getSelfPid

  forM_ processes $ \process -> monitor process
  -- Send tasks to the slave nodes
  say "master started"
  sendTasks processes mypid

expectLoop :: Process ()
expectLoop = do
  e <- expect
  say e

findSlaves :: Process [NodeId]
findSlaves = do
  -- Discover and list available slave nodes (you may need to replace these with actual slave node addresses)
  -- For simplicity, we assume that slave nodes are running on the same machine.
  return $ map NodeId [EndPointAddress . BS.pack $ "localhost:10501", EndPointAddress . BS.pack $ "localhost:10502"]

getProcess :: [NodeId] -> Process [ProcessId]
getProcess = do
  traverse (\slave -> spawn slave $ $(mkClosure 'slaveProcess) ("" :: String, 0 :: Int))

sendTasks :: [ProcessId] -> ProcessId -> Process ()
sendTasks processes mypid = do
  -- Distribute tasks to slaves
  let task1 = [(1, 2), (3, 4)] :: [(Int, Int)] -- Tasks to distribute
  let task2 = [("1", "2"), ("3", "4")] :: [(String, String)] -- Tasks to distribute
  forM_ (zip3 task1 task2 processes) $ \(taskInt, taskStr, process) -> do
    send process $ IntMsg taskInt
    send process $ StringMsg taskStr
  say "master started"
  a <- forM processes $ (\process -> query process mypid)
  say "master started"
  expectLoop
  -- forM a $ \(s1, i1) -> do 
  --   say s1
  --   say . show $ i1
  -- return ()

type Host = String
type Port = String
getTransport :: Host -> Port -> IO Transport
getTransport host port = do
  Right transport <- flip createTransport defaultTCPParameters $ defaultTCPAddr host port
  return transport

main :: IO ()
main = do
  print "start"
  master_transport <- getTransport "localhost" "10500"
  slave1_transport <- getTransport "localhost" "10501"
  slave2_transport <- getTransport "localhost" "10502"
  print "start"
  
  slave_local_nodes <- flip traverse [slave1_transport, slave2_transport] (\tr -> newLocalNode tr myRemoteTable)
  print "start"
  let slave_nodes = map localNodeId slave_local_nodes
  print "start"
  slaves_process <- forM slave_local_nodes (\tr -> 
    forkProcess tr (receiveWait[]))
  print "slave in work"
  node <- newLocalNode master_transport myRemoteTable
  runProcess node $ master



app/Main slave.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}


import qualified Data.ByteString.Char8 as BS
import System.Environment (getArgs)
import GHC.Generics
import Control.Monad (forM, forM_)
import Data.Typeable
import Data.Binary (Binary)
import Control.Concurrent (threadDelay)
import Control.Monad (forever)
import Control.Distributed.Process
import Control.Distributed.Process.Closure
import Control.Distributed.Process.Node
import Control.Distributed.Process.Serializable
import Network.Transport.TCP (createTransport, defaultTCPParameters)
import Network.Transport (Transport, EndPointAddress(..))
import Control.Concurrent (forkIO)
import Control.Distributed.Process.Backend.SimpleLocalnet

replyBack :: (ProcessId, String) -> Process ()
replyBack (sender, msg) = send sender msg

logMessage :: String -> Process ()
logMessage msg = say $ "handling " ++ msg


data MyMessage = StringMsg (String, String) |
               IntMsg (Int, Int) |
               QueryMsg ProcessId |
               ResultMsg (String, Int) |
               TerminateMsg deriving (Generic, Binary)

data (Show a) => Result a = Result { _getResult :: a }
instance (Show a) => Show (Result a) where
  show a = show $ _getResult a

showResult :: (Show a) => Result a -> Process ()
showResult a = say $ show a

addNumbers :: (Int, Int) -> Int
addNumbers (x, y) = (x + y)

stringTransform :: String -> String
stringTransform str = reverse str

stringTransformWrapper :: (String, String) -> String
stringTransformWrapper (one, two) = stringTransform $ show one <> show two

slaveProcess :: (String, Int) -> Process ()
slaveProcess (s, i) = do
  say $ s <> show i
  (s', i') <- receiveWait[match commandCatch]
  say $ s' <> show i'
  slaveProcess (s', i')
  where
    commandCatch (QueryMsg pid) = do
      send pid (s, i)
      return (s, i)
    commandCatch (StringMsg (s1, s2)) = return ((s <> (stringTransformWrapper (s1, s2))), i)
    commandCatch (IntMsg (i1, i2)) = return (s, (i + (addNumbers (i1, i2))))
    commandCatch TerminateMsg = terminate

query :: ProcessId -> ProcessId -> Process ()
query cpid mypid = do
  send cpid (QueryMsg mypid)


remotable ['slaveProcess]

myRemoteTable :: RemoteTable
myRemoteTable = Main.__remoteTable initRemoteTable

master :: Backend -> [NodeId] -> Process ()
master backend slaves = do
  say $ "Slaves: " ++ show slaves
  say "master started"
  -- Get information about available slave nodes
  -- slaves <- findSlaves
  -- say "master started"

  -- processes <- getProcess slaves
  -- say "master started"
  -- mypid <- getSelfPid

  -- forM_ processes $ \process -> send process "hello" >> send process (mypid, "hello")
  -- forM_ processes $ \process -> monitor process
  -- forM_ slaves $ \process -> monitorNode process
  -- -- Send tasks to the slave nodes
  -- say "master started"
  -- sendTasks processes mypid

expectLoop :: Process ()
expectLoop = do
  e <- expect
  say e

-- findSlaves :: Process [NodeId]
-- findSlaves = do
--   -- Discover and list available slave nodes (you may need to replace these with actual slave node addresses)
--   -- For simplicity, we assume that slave nodes are running on the same machine.
--   return $ map NodeId [EndPointAddress . BS.pack $ "localhost:10501", EndPointAddress . BS.pack $ "localhost:10502"]

-- getProcess :: [NodeId] -> Process [ProcessId]
-- getProcess = do
--   traverse (\slave -> spawn slave $ $(mkClosure 'slaveProcess) ("" :: String, 0 :: Int))

-- sendTasks :: [ProcessId] -> ProcessId -> Process ()
-- sendTasks processes mypid = do
--   -- Distribute tasks to slaves
--   let task1 = [(1, 2), (3, 4)] :: [(Int, Int)] -- Tasks to distribute
--   let task2 = [("1", "2"), ("3", "4")] :: [(String, String)] -- Tasks to distribute
--   forM_ (zip3 task1 task2 processes) $ \(taskInt, taskStr, process) -> do
--     send process $ IntMsg taskInt
--     send process $ StringMsg taskStr
--   say "master started"
--   a <- forM processes $ (\process -> query process mypid)
--   say "master started"
--   expectLoop
  -- forM a $ \(s1, i1) -> do 
  --   say s1
  --   say . show $ i1
  -- return ()


-- type Host = String
-- type Port = String
-- getTransport :: Host -> Port -> IO Transport
-- getTransport host port = do
--   Right transport <- flip createTransport defaultTCPParameters $ defaultTCPAddr host port
--   return transport


main :: IO ()
main = do
  args <- getArgs

  case args of
    ["master", host, port] -> do
      backend <- initializeBackend host port myRemoteTable
      startMaster backend (master backend)
    ["slave", host, port] -> do
      backend <- initializeBackend host port myRemoteTable
      startSlave backend


-- main :: IO ()
-- main = do
--   args <- getArgs

--   case args of
--     ["master", host, port] -> do
--         master_transport <- getTransport host port
--         node <- newLocalNode master_transport myRemoteTable
--         runProcess node $ master
--     ["slave", host, port] -> do
--         master_transport <- getTransport host port
--         node <- newLocalNode master_transport myRemoteTable
--         runProcess node $ (receiveWait[match logMessage, match replyBack])


-- main :: IO ()
-- main = do
--   print "start"
--   master_transport <- getTransport "localhost" "10500"
--   slave1_transport <- getTransport "localhost" "10501"
--   slave2_transport <- getTransport "localhost" "10502"
--   print "start"
  
--   slave_local_nodes <- flip traverse [slave1_transport, slave2_transport] (\tr -> newLocalNode tr myRemoteTable)
--   print "start"
--   let slave_nodes = map localNodeId slave_local_nodes
--   print "start"
--   slaves_process <- forM slave_local_nodes (\tr -> 
--     forkProcess tr (receiveWait[]))
--   print "slave in work"
--   node <- newLocalNode master_transport myRemoteTable
--   runProcess node $ master



app/Main.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

import qualified Data.ByteString.Char8 as BS
import System.Environment (getArgs)
import GHC.Generics
import Control.Monad (forM, forM_)
import Data.Typeable
import Data.Binary (Binary)
import Control.Concurrent (threadDelay)
import Control.Monad (forever)
import Control.Distributed.Process
import Control.Distributed.Process.Closure
import Control.Distributed.Process.Node
import Control.Distributed.Process.Serializable
import Network.Transport.TCP (createTransport, defaultTCPParameters, defaultTCPAddr)
import Network.Transport (Transport, EndPointAddress(..))
import Control.Concurrent (forkIO)


replyBack :: (ProcessId, String) -> Process ()
replyBack (sender, msg) = send sender msg

logMessage :: String -> Process ()
logMessage msg = say $ "handling " ++ msg


data MyMessage = StringMsg (String, String) |
               IntMsg (Int, Int) |
               QueryMsg ProcessId |
               ResultMsg (String, Int) |
               TerminateMsg deriving (Generic, Binary)

data (Show a) => Result a = Result { _getResult :: a }
instance (Show a) => Show (Result a) where
  show a = show $ _getResult a

showResult :: (Show a) => Result a -> Process ()
showResult a = say $ show a

addNumbers :: (Int, Int) -> Int
addNumbers (x, y) = (x + y)

stringTransform :: String -> String
stringTransform str = reverse str

stringTransformWrapper :: (String, String) -> String
stringTransformWrapper (one, two) = stringTransform $ show one ++ show two

slaveProcess :: (String, Int) -> Process ()
slaveProcess (s, i) = do
  say $ "before: string: " <> s <> "and number:" <> show i
  (s', i') <- receiveWait[match commandCatch]
  say $ "after: string: " <> s' <> "and number:" <> show i'
  slaveProcess (s', i')
  where
    commandCatch (QueryMsg pid) = do
      send pid (s, i)
      a <- getProcessInfo pid
      say $ show a
      say $ "send to master"
      return (s, i)
    commandCatch (StringMsg (s1, s2)) = return (id (s <> (stringTransformWrapper (s1, s2))), i)
    commandCatch (IntMsg (i1, i2)) = return (s, (i + (addNumbers (i1, i2))))
    commandCatch TerminateMsg = do
      say "died"
      terminate

query :: ProcessId -> ProcessId -> Process (String, Int)
query cpid mypid = do
  send cpid (QueryMsg mypid)
  expect


remotable ['slaveProcess]

myRemoteTable :: RemoteTable
myRemoteTable = Main.__remoteTable initRemoteTable

getStats :: NodeId -> Process ()
getStats n = do
  b <- getNodeStats n
  say $ show b

master :: Process ()
master = do
  say "master started"
  -- Get information about available slave nodes
  slaves <- findSlaves
  say "looking for slaves"
  forM_ slaves getStats 
  say $ show slaves
  mypid <- getSelfPid
  -- forM_ slaves $ \process -> send process "hello" >> send process (mypid, "hello")
  -- forM_ slaves $ \process -> monitorNode process
  -- say "say hello!"
  processes <- getProcess slaves
  forM_ processes $ \process -> monitor process
  say "start processes on slaves"
  -- Send tasks to the slave nodes
  say "send tasks"
  sendTasks processes mypid

expectLoop :: Process ()
expectLoop = do
  e <- expect
  say e

findSlaves :: Process [NodeId]
findSlaves = do
  -- Discover and list available slave nodes (you may need to replace these with actual slave node addresses)
  -- For simplicity, we assume that slave nodes are running on the same machine.
  return $ map NodeId [EndPointAddress . BS.pack $ "127.0.0.1:8081:0", EndPointAddress . BS.pack $ "127.0.0.1:8080:0"]

getProcess :: [NodeId] -> Process [ProcessId]
getProcess = do
  traverse (\slave -> spawn slave $ $(mkClosure 'slaveProcess) ("" :: String, 0 :: Int))

sendTasks :: [ProcessId] -> ProcessId -> Process ()
sendTasks processes mypid = do
  -- Distribute tasks to slaves
  let task1 = [(1, 2), (3, 4)] :: [(Int, Int)] -- Tasks to distribute
  let task2 = [("1", "2"), ("3", "4")] :: [(String, String)] -- Tasks to distribute
  forM_ (zip3 task1 task2 processes) $ \(taskInt, taskStr, process) -> do
    send process $ IntMsg taskInt
    send process $ StringMsg taskStr
  a <- forM processes $ (\process -> query process mypid)
  say $ "get results"
  forM_ a (\a -> say $ "string: " <> fst a <> "and number:" <> show (snd a))
  forM_ processes $ (\process -> send process (TerminateMsg))
  let count = 0 :: Int
  forever $ do
    say "start waiting"
    reason <- expect :: Process ProcessMonitorNotification
    say $ show reason
    -- liftIO $ threadDelay 1000000
    -- let count = count + 1
    -- if (count == 2) then terminate
    -- else (say "hayo") 




type Host = String
type Port = String
getTransport :: Host -> Port -> IO Transport
getTransport host port = do
  Right transport <- flip createTransport defaultTCPParameters $ defaultTCPAddr host port
  return transport


main :: IO ()
main = do
  args <- getArgs

  case args of
    ["master", host, port] -> do
        master_transport <- getTransport host port
        node <- newLocalNode master_transport myRemoteTable
        runProcess node $ master
    ["slave", host, port] -> do
        master_transport <- getTransport host port
        node <- newLocalNode master_transport myRemoteTable
        runProcess node $ do
          a <- getLocalNodeStats 
          say $ show a
          receiveWait[match logMessage, match replyBack]


-- main :: IO ()
-- main = do
--   print "start"
--   master_transport <- getTransport "localhost" "10500"
--   slave1_transport <- getTransport "localhost" "10501"
--   slave2_transport <- getTransport "localhost" "10502"
--   print "start"
  
--   slave_local_nodes <- flip traverse [slave1_transport, slave2_transport] (\tr -> newLocalNode tr myRemoteTable)
--   print "start"
--   let slave_nodes = map localNodeId slave_local_nodes
--   print "start"
--   slaves_process <- forM slave_local_nodes (\tr -> 
--     forkProcess tr (receiveWait[]))
--   print "slave in work"
--   node <- newLocalNode master_transport myRemoteTable
--   runProcess node $ master

